Split sentences_detailed.csv into single language files

AndiPersti opened the pull request:
As discussed on #1729 here's one possible way to serve single language exports.

I don't think it's necessary to split both `sentences.csv` and `sentences_detailed.csv` so I've chosen the latter because it is a superset of the former.

Even if we decide to not use this implementation we should still change line 7 and line 10:
- `ROOT` is defined on line 5 so we should use it on line 7
- `var/tmp` is also used by `systemd` so we should just move the `csv` files we've created

jiru commented:
Thank you, Andi. :smile_cat: 

A few comments.

* The wording `single_language` feels not self-explicit enough. How about `per_language`?
* You should update the Downloads page to give some visibility to the new files.
* If we are to provide per-language versions of other existing CSV exports, it may be a good idea to group them into per-language directories, so we’d have something like `exports/*.csv` for the current files, and `exports/per_language/deu/*.csv`, `exports/per_language/fra/*.csv` etc. for the per-language ones. This way, since the per-language files have the same format, we’d only need to put an additional link to the `exports/per_language/` directory on the Downloads page, which will give a show of language codes one can choose from.

AndiPersti commented:
>The wording `single_language` feels not self-explicit enough. How about `per_language`?

Yes, `single_language` also sounded a little bit strange to me but I couldn't find a better name.

>You should update the Downloads page to give some visibility to the new files.

I've added a note about them under the `Sentences` and `Sentences (CC0)` sections but I'm not sure about the placement and the wording.
(BTW: There's text in the code which doesn't show up on the page because it misses the `echo` call. Is it still useful: https://github.com/Tatoeba/tatoeba2/blob/2f40784c100bdf5cfec850865cb7d6bbec10599b/src/Template/Pages/downloads.ctp#L199-L202 )

>If we are to provide per-language versions of other existing CSV exports, it may be a good idea to group them into per-language directories, so we’d have something like `exports/*.csv` for the current files, and `exports/per_language/deu/*.csv`, `exports/per_language/fra/*.csv` etc. for the per-language ones.

I've changed the script to split all three `sentences` files and put the per-language files into their corresponding subdirectories.

jiru commented on src/Template/Pages/downloads.ctp:
```diff
@@ -146,6 +146,9 @@ $this->set('title_for_layout', $this->Pages->formatTitle(__('Download sentences'
         $tab_str = __('tab');
+        $per_language_str = __('Per-language files are available at <a ' .
+            'href="https://downloads.tatoeba.org/exports/per_language/">' .
+            'https://downloads.tatoeba.org/exports/per_language</a>.');
```
There are two problems with this translatable string.

1. It includes a URL, however the URL shouldn’t be showed to translators because they may modify it by mistake. Instead, use:
```php
$per_language_str = format(
    __('Here is my <a href="{}">link</a>.'),
    'https://downloads.tatoeba.org/exports/per_language'
);
```

2. The [hyperlink text is not good](https://agsci.psu.edu/communications/how-to/websites/web-best-practices/appropriate-link-text). I know that other existing links on that page are like that too, but feel free to make it right.

AndiPersti replied:
> I know that other existing links on that page are like that too, but feel free to make it right.

How about something like this:
![downloads](https://user-images.githubusercontent.com/6566321/75660772-7527fe00-5c6c-11ea-9ff9-7236a4c64cb2.png)

trang replied:
You could have a dropdown to select the language and have a download button.

![Screenshot_2020-03-03 Download sentences - Tatoeba](https://user-images.githubusercontent.com/221850/75809596-51a6a580-5d89-11ea-8ecf-040e6243859b.png)

For the files where there's no possibility to download in a specific language, you just don't have the dropdown.

jiru replied:
The language drop down seems like a better idea. I think it’s important however to let people do "right-click → save as" or "right-click → copy link URL", so the download button should allow to do so. I think you could add a checkbox that makes the language dropdown selectable, like this:

> [✘] Only include sentences in: [Select a language]
> [Download]

This way, it makes it easier to understand that the language dropdown is a filtering option.

@AndiPersti Or, if you don’t feel like doing all that frontend stuff, make it simple:

> <h3>Sentences</h3>
>
> <b>Filename</b>
> sentences.csv
>
> <b>Download</b>
> Download this file <a href="https://downloads.tatoeba.org/exports/sentences.tar.bz2">including all languages</a> or <a href="https://downloads.tatoeba.org/exports/per_language/">only for a single language</a>.

Or

> Download this file <a href="http://example.com/">including all languages</a>.

AndiPersti commented:
I'm sorry for missing this problem and try to fix the error tomorrow.

AndiPersti commented:
Some screenshots:
![downloads](https://user-images.githubusercontent.com/6566321/76282187-5393cb80-6297-11ea-8bb2-89230367bae2.png)
![downloads2](https://user-images.githubusercontent.com/6566321/76282197-57bfe900-6297-11ea-93b1-4f70a6c3970f.png)

jiru commented on src/View/Helper/DownloadsHelper.php:
```diff
@@ -0,0 +1,121 @@
+    public $helpers = ['Html', 'Languages'];
+
+    const PER_LANGUAGE_DIR = '/var/www-downloads/exports/per_language';
+    const DOWNLOAD_URL = 'https://downloads.tatoeba.org/exports';
```
We have the habit of putting this kind of constants in `app_local.php`, so that it’s easier to integrate with Imouto or any other custom instance. If you go for this, please add new variables in `app_local.php.template` that can be readily used by Imouto. Within your tests, you can then add the appropriate `Configure::write()` calls in the `setUp()` method.

jiru commented on src/View/Helper/DownloadsHelper.php:
```diff
@@ -0,0 +1,121 @@
+        $command = 'find ' . self::PER_LANGUAGE_DIR . ' -type f ' .
+                   "-name '*$basename.tsv.bz2' -printf '%P\\n' " .
+                   '2> /dev/null';
+        $stdout = trim(shell_exec($command));
```
Executing shell commands within a PHP script is not a good practice in terms of security. If `$basename` is tainted, it allows all sorts of shell injection. Please consider rewriting this part in [pure PHP](https://book.cakephp.org/3/en/core-libraries/file-folder.html) instead.

jiru commented on src/View/Helper/DownloadsHelper.php:
```diff
@@ -0,0 +1,121 @@
+        if (!empty($urls)) {
+            $languages = array_intersect_key(
+                $this->Languages->onlyLanguagesArray(false) +
+                ['unknown' => __x('dropdown-list', 'Unknown language')],
```
Feel free to make `unknownLanguagesArray()` accept a parameter and use `$this->Languages->unknownLanguagesArray(false)`.

AndiPersti replied:
For my very first prototype I've constructed the array myself and `unknown` was in between the other languages after localized sorting.

Then I've looked around and found the methods in the `Languages` helper. But in addition to the splitting problem, `unknowLanguagesArray()` also puts `unknown` at the beginning which I personally don't like very much (after loading the page it would show the two options `All languages` and `Only sentences in Unknown language`).

So I've used the current version which puts `unknown` at the end.

Where do you think `unknown` should be put: at the beginning, at the end or in between?

jiru replied:
The only place `unknownLanguagesArray()` is used is the language selector on `/sentences/show_all_in`. I think it’s reasonable to put it at the end because it’s less frequently used than other languages.

jiru commented on docs/cron/export.sh:
```diff
@@ -4,10 +4,10 @@ set -e
-mysql -u "$DB_USER" -p"$DB_PASS" "$DB" < /var/www-prod/docs/database/scripts/weekly_exports.sql
+mysql -u "$DB_USER" -p"$DB_PASS" "$DB" < $ROOT/docs/database/scripts/weekly_exports.sql
 
 DL_DIR="/var/www-downloads/exports"
```
If you make this a variable in `app_local.php`, you should also remove this line and edit the cron job runner `docs/cron/runner.sh` so that it passes the variable to this script instead.

jiru commented on src/Template/Element/per_language_files.ctp:
```diff
@@ -0,0 +1,51 @@
+                    <?php endforeach; ?>
+                </md-select>
+            </div>
+        <md-radio-group>
```
This should be a closing tag instead.

jiru commented:
Wow, it looks good! Thank you for taking the time to reformat the contents of download.ctp, it’s so much cleaner now. :clap: 

If you don’t feel like making `/var/www-downloads/exports/` and `https://downloads.tatoeba.org/exports` configurable, no problem, I can still merge your PR without that, but the other changes I asked are rather important.

It would be great if I could merge your PR by Wednesday 11th because it introduces some translatable string changes. If not, it’s okay because the string changes are minor anyway.

AndiPersti commented:
> It would be great if I could merge your PR by Wednesday 11th because it introduces some translatable string changes. If not, it’s okay because the string changes are minor anyway.

I'll try my best but I'm not sure if I will have enough time tomorrow.

AndiPersti commented:
I've implemented all requested changes and rebased the branch onto the current dev branch because there was a merge conflict in `webroot/js/responsive/app.module.js`.

jiru commented:
Thank you very much, Andreas! :+1:
You’re doing such a good job, it’s awesome.

alanfgh commented:
Many thanks for this. As a result of this change, and probably some previous ones that got rid of nulls and other bad characters, I can now search through all the English sentences with a text editor without any problem. Very nice!
